### PR TITLE
Add testing platform architecture

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Move and reorganize the architecture/extensibility documentation for the new testing platform.

cc @MarcoRossignoli 

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/testing/unit-testing-platform-architecture-capabilities.md](https://github.com/dotnet/docs/blob/2eb27ca02f9851d0cc2f43fd0a7618fc7eb72f67/docs/core/testing/unit-testing-platform-architecture-capabilities.md) | [docs/core/testing/unit-testing-platform-architecture-capabilities](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-architecture-capabilities?branch=pr-en-us-41735) |
| [docs/core/testing/unit-testing-platform-architecture-extensions.md](https://github.com/dotnet/docs/blob/2eb27ca02f9851d0cc2f43fd0a7618fc7eb72f67/docs/core/testing/unit-testing-platform-architecture-extensions.md) | [docs/core/testing/unit-testing-platform-architecture-extensions](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-architecture-extensions?branch=pr-en-us-41735) |
| [docs/core/testing/unit-testing-platform-architecture-services.md](https://github.com/dotnet/docs/blob/2eb27ca02f9851d0cc2f43fd0a7618fc7eb72f67/docs/core/testing/unit-testing-platform-architecture-services.md) | [docs/core/testing/unit-testing-platform-architecture-services](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-architecture-services?branch=pr-en-us-41735) |
| [docs/core/testing/unit-testing-platform-architecture.md](https://github.com/dotnet/docs/blob/2eb27ca02f9851d0cc2f43fd0a7618fc7eb72f67/docs/core/testing/unit-testing-platform-architecture.md) | [docs/core/testing/unit-testing-platform-architecture](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-architecture?branch=pr-en-us-41735) |
| [docs/core/testing/unit-testing-platform-exit-codes.md](https://github.com/dotnet/docs/blob/2eb27ca02f9851d0cc2f43fd0a7618fc7eb72f67/docs/core/testing/unit-testing-platform-exit-codes.md) | [docs/core/testing/unit-testing-platform-exit-codes](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-exit-codes?branch=pr-en-us-41735) |
| [docs/core/testing/unit-testing-platform-intro.md](https://github.com/dotnet/docs/blob/2eb27ca02f9851d0cc2f43fd0a7618fc7eb72f67/docs/core/testing/unit-testing-platform-intro.md) | [docs/core/testing/unit-testing-platform-intro](https://review.learn.microsoft.com/en-us/dotnet/core/testing/unit-testing-platform-intro?branch=pr-en-us-41735) |

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
I don't see the image and the mermaid rendering correctly is it related only to the "preview"?

**Comment by Evangelink** [comment]
@IEvangelist Are our docs supporting mermaid?

**Review by IEvangelist — COMMENTED** [review]
I'm still not done with this review, it's taking a really long time - been working on it for a while now. I'm going to comment on it, and return when I can focus a bit more. But for now, here's what I have.

File: docs/core/testing/media/bus.png
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/media/bus.png:1]: Any extension to the new platform (including Test Framework) can publish or subscribe to data from the message bus.

[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/media/bus.png:1]: I'll suggest a different design based on the tool you suggested, thanks!

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 1]
```diff
@@ -0,0 +1,1164 @@
...
+```
+
+> [!NOTE]
+> Returning empty [ITestFrameworkCapabilities](./unit-testing-platform-architecture-capabilities.md) should not prevent the execution of the test session. All test frameworks should be capable of discovering and running tests. The impact should be limited to extensions that may opt out if the test framework lacks a certain feature.
+
+### Creating a testing framework
+
+The `Microsoft.Testing.Platform.Extensions.TestFramework.ITestFramework` is implemented by extensions that provide a test framework:
+
+```cs
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-platform-architecture-extensions.md]: I have updated all remaining instances! Thanks!

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 2]
```diff
@@ -0,0 +1,1164 @@
...
+```
+
+Entry point with the registration:
+
+```cs
+var testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
+// Register the testing framework
+testApplicationBuilder.AddTestingFramework();
+using var testApplication = await testApplicationBuilder.BuildAsync();
+return await testApplication.RunAsync();
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-platform-architecture-extensions.md:138]: We internally enforce the same rules as runtime for `var` but some examples were created from different projects so we indeed end up with multiple styles. For docs, I am fine with both directions, please let me know what you prefer (maybe explicit is better for docs?).

> Also, what are we returning here? Is this a top-level program, and we return a Task<int>?

Correct, that's a top-level program returning `Task<int>` with the exit code of the test application.

File: docs/core/testing/media/bus.png
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/media/bus.png:1]: cc @MarcoRossignoli

**Comment by Evangelink** [comment]
> I'm still not done with this review, it's taking a really long time - been working on it for a while now. 

Thank you so much for everything you have reviewed already! We know this is a super big change, I have read it a few times and I am still not 100% happy with the results of what we wrote nor the structure but it's not always easy to get the abstraction of knowing all the implementation details.

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 3]
```diff
@@ -0,0 +1,1160 @@
...
+
+The `ITestSessionLifeTimeHandler` is an *in-process* extension that enables the execution of code *before* and *after* the test session.
+
+To register a custom `ITestSessionLifeTimeHandler`, utilize the following API:
+
+```csharp
+ITestApplicationBuilder testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
+...
+testApplicationBuilder.TestHost.AddTestSessionLifetimeHandle(serviceProvider => new CustomTestSessionLifeTimeHandler());
+...
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:760]:
```suggestion
ITestApplicationBuilder builder = await TestApplication.CreateBuilderAsync(args);

// ...

builder.TestHost.AddTestSessionLifetimeHandle(
    static serviceProvider => new CustomTestSessionLifeTimeHandler());
```

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 4]
```diff
@@ -0,0 +1,1160 @@
...
+}
+
+public interface ITestHostExtension : IExtension
+{
+}
+```
+
+The `ITestSessionLifetimeHandler` is a type of `ITestHostExtension`, which serves as a base for all *test host* extensions. Like all other extension points, it also inherits from [IExtension](#the-iextension-interface). Therefore, like any other extension, you can choose to enable or disable it using the `IExtension.IsEnabledAsync` API.
+
+Let's describe the api:
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:789]:
```suggestion
Consider the following details for this API:
```

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 5]
```diff
@@ -0,0 +1,1160 @@
...
+*This extension point is crucial as it enables developers to gather and process all the information generated during a test session.*
+
+To register a custom `IDataConsumer`, utilize the following api:
+
+```csharp
+ITestApplicationBuilder testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
+...
+testApplicationBuilder.TestHost.AddDataConsumer(serviceProvider
+    => new CustomDataConsumer());
+...
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:855]:
```suggestion
ITestApplicationBuilder builder = await TestApplication.CreateBuilderAsync(args);

// ...

builder.TestHost.AddDataConsumer(
    static serviceProvider => new CustomDataConsumer());
```

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 6]
```diff
@@ -0,0 +1,1160 @@
...
+            ...
+        }
+
+        return Task.CompletedTask;
+    }
+...
+}
+```
+
+Finally, the api takes a `CancellationToken` which the extension is expected to honor.
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:929]: API should be always be uppercase.

```suggestion
Finally, the API takes a `CancellationToken` which the extension is expected to honor.
```

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 7]
```diff
@@ -0,0 +1,1160 @@
...
+The `ITestHostEnvironmentVariableProvider` is an *out-of-process* extension that enables you to establish custom environment variables for the test host. Utilizing this extension point ensures that the testing platform will initiate a new host with the appropriate environment variables, as detailed in the [architecture](./unit-testing-platform-architecture.md) section.
+
+To register a custom `ITestHostEnvironmentVariableProvider`, utilize the following API:
+
+```csharp
+ITestApplicationBuilder testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
+...
+testApplicationBuilder.TestHostControllers.AddEnvironmentVariableProvider(serviceProvider =>
+    => new CustomEnvironmentVariableForTestHost());
+...
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:952]:
```suggestion
ITestApplicationBuilder builder = await TestApplication.CreateBuilderAsync(args);

// ...

builder.TestHostControllers.AddEnvironmentVariableProvider(
    static serviceProvider => new CustomEnvironmentVariableForTestHost());
```

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 8]
```diff
@@ -0,0 +1,1160 @@
...
+    public string Variable { get; }
+    public string? Value { get; }
+    public bool IsSecret { get; }
+    public bool IsLocked { get; }
+}
+```
+
+The `ITestHostEnvironmentVariableProvider` is a type of `ITestHostControllersExtension`, which serves as a base for all *test host controller* extensions. Like all other extension points, it also inherits from [IExtension](#the-iextension-interface). Therefore, like any other extension, you can choose to enable or disable it using the `IExtension.IsEnabledAsync` API.
+
+Let's describe the api:
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:997]:
```suggestion
Consider the details for this API:
```

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 9]
```diff
@@ -0,0 +1,1160 @@
...
+1. [ITestFramework.ExecuteRequestAsync](#test-framework-extension): In-process, this method can be called one or more times. At this point, the testing framework will transmit information to the [IMessageBus](./unit-testing-platform-architecture-services.md#the-imessagebus-service) that can be utilized by the [IDataConsumer](#the-idataconsumer-extensions).
+1. [ITestFramework.CloseTestSessionAsync](#test-framework-extension): In-process
+1. [ITestSessionLifetimeHandler.OnTestSessionFinishingAsync](#the-itestsessionlifetimehandler-extensions): In-process
+1. [ITestApplicationLifecycleCallbacks.AfterRunAsync](#the-itestsessionlifetimehandler-extensions): In-process
+1. In-process cleanup, involves calling dispose and [IAsyncCleanableExtension](#asynchronous-initialization-and-cleanup-of-extensions) on all extension points.
+1. [ITestHostProcessLifetimeHandler.OnTestHostProcessExitedAsync](#the-itestsessionlifetimehandler-extensions) : Out-of-process
+1. Out-of-process cleanup, involves calling dispose and [IAsyncCleanableExtension](#asynchronous-initialization-and-cleanup-of-extensions) on all extension points.
+
+## Extensions helpers
+
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:1091]: We need a sentence to tie these headings together.

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 10]
```diff
@@ -0,0 +1,1160 @@
...
+Once you've created the `CompositeExtensionFactory<CustomExtension>` for your type, you can register it with both the `IDataConsumer` and `ITestSessionLifetimeHandler` APIs, which offer an overload for the `CompositeExtensionFactory<T>`:
+
+```csharp
+ITestApplicationBuilder testApplicationBuilder = await TestApplication.CreateBuilderAsync(args);
+...
+CompositeExtensionFactory<CustomExtension> compositeExtensionFactory = new(serviceProvider => new CustomExtension());
+testApplicationBuilder.TestHost.AddTestSessionLifetimeHandle(compositeExtensionFactory);
+testApplicationBuilder.TestHost.AddDataConsumer(compositeExtensionFactory);
+...
```
Comment (IEvangelist) [inline docs/core/testing/unit-testing-platform-architecture-extensions.md:1148]:
```suggestion
ITestApplicationBuilder builder = await TestApplication.CreateBuilderAsync(args);

// ...

CompositeExtensionFactory<CustomExtension> factory = new(serviceProvider => new CustomExtension());

builder.TestHost.AddTestSessionLifetimeHandle(factory);

builder.TestHost.AddDataConsumer(factory);
```

**Review by IEvangelist — COMMENTED** [review]
There are still a few things that we'll need to update, this is starting to shape up though. Let's consider this 2/3 in terms of reviews. 🤓

**Comment by IEvangelist** [comment]
I've checked back on this one a few times, there's still a lot of suggestions that haven't been resolved. I'll come back once those are addressed. /cc @Evangelink

**Comment by Evangelink** [comment]
@IEvangelist Yes sorry! I have been pretty busy with other priorities, I'll get back here early next week!

**Comment by IEvangelist** [comment]
> @IEvangelist Yes sorry! I have been pretty busy with other priorities, I'll get back here early next week!

No worries, I completely understand. I just wanted to make sure I wasn't missing anything. I'm off all next week, so I'll check back the week of 8/12. Thank you, my friend! @Evangelink

File: docs/core/testing/unit-testing-platform-architecture-extensions.md [hunk 11]
```diff
@@ -0,0 +1,1164 @@
...
+```cs
+public interface ITestFramework : IExtension
+{
+    Task<CreateTestSessionResult> CreateTestSessionAsync(CreateTestSessionContext context);
+    Task ExecuteRequestAsync(ExecuteRequestContext context);
+    Task<CloseTestSessionResult> CloseTestSessionAsync(CloseTestSessionContext context);
+}
+```
+
+The `ITestFramework` interface inherits from the [IExtension](#the-iextension-interface) interface, which is an interface that all extension points inherit from. `IExtension` is used to retrieve the name and description of the extension. The `IExtension` also provides a way to dynamically enable or disable the extension in setup, through `Task<bool> IsEnabledAsync()`, please make sure that you return `true` from this method if you have no special needs.
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply docs/core/testing/unit-testing-platform-architecture-extensions.md]: Thanks David! This was considered but we decided to favor `Task` as `ValueTask` isn't supported for all tfms we support. Having conditional compilation is possible but was causing some other troubles. We added a ticket to revise in v2.

**Comment by Evangelink** [comment]
Hey @IEvangelist, could we resume work on this PR? We have finished integration with xUnit v3, NUnit, and we are getting close to finishing with F# Expecto.

**Comment by Evangelink** [comment]
@IEvangelist Could we resume work on this PR when you have some time? I think I have handled all the comments you made to me.

**Comment by IEvangelist** [comment]
We need to address the build warnings, [see here](https://buildapi.docs.microsoft.com/Output/PullRequest/a3bda507-3390-de91-8f7f-26f90f4e5fc8/202410030930004253-41735/BuildReport?accessString=3713538ddc689be483e16f94ea69f2647a4b80e6c13b0649d1ba6883e46185a0). Then, let's ensure that all the images are cropped nicely, then we should be able to merge.

**Comment by Evangelink** [comment]
@IEvangelist PR is green :) A HUGE thank you for your support through this massive update.

**Review by IEvangelist — APPROVED** [review]

**Pull request merged**